### PR TITLE
Expose AdapterSpecification through AdapterContext.

### PR DESCRIPTION
**Pull request by chtyim**
- Define the AdapterSpecification interface in cdap-api
- Make AdapterDefinition to implement AdapterSpecification
- Also piggy back to fix a small bug in DefaultAdapterConfigurer for wrong check on plugin usage.

### CODE REVIEW

File: cdap-api/src/main/java/co/cask/cdap/api/templates/AdapterSpecification.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+  /**
+   * Returns the name of the adapter.
+   */
+  String getName();
+
+  /**
+   * Returns the description of the adapter.
+   */
+  String getDescription();
```
Comment (bdmogal) [inline cdap-api/src/main/java/co/cask/cdap/api/templates/AdapterSpecification.java]: The description can never be null, right?

File: cdap-app-fabric/src/main/java/co/cask/cdap/templates/AdapterDefinition.java [hunk 2]
```diff
@@ -145,6 +152,16 @@ public JsonElement getConfig() {
...
   }
 
+  @Override
+  public <T> T getConfig(Type configType) {
+    return GSON.fromJson(config, configType);
+  }
+
+  @Override
+  public String getConfigString() {
+    return config.toString();
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/templates/AdapterDefinition.java]: This is the one that extends ``PluginConfig``, right? Its ``toString()`` returns a JSON?

Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/templates/AdapterDefinition.java]: config is of type `JsonElement`, not the user defined class.

**Comment by bdmogal** [comment]
:+1:

**Pull request merged**